### DISCUSSION TITLE
Splitstore discussion

### DISCUSSION THREAD

**Discussion by arajasek** [filecoin-project/lotus#5788] [Show and tell]
Lotus 1.5.1 introduced an experimental splitstore in #4992. This feature reduces the performance impact of the very large Filecoin datastore.

## Design

This release also introduces the splitstore, a new optional blockstore that segregates the monolithic blockstore into cold and hot regions. The hot region contains objects from the last 4-5 finalities plus all reachable objects from two finalities away. All other objects are moved to the cold region using a compaction process that executes every finality, once 5 finalities have elapsed.

The splitstore allows us to separate the two regions quite effectively, using two separate badger blockstores. The separation means that the live working set is much smaller, which results in potentially significant performance improvements. In addition, it means that the coldstore can be moved to a separate (bigger, slower, cheaper) disk without loss of performance.

The design also allows us to use different implementations for the two blockstores; for example, an append-only blockstore could be used for coldstore and a faster memory mapped blockstore could be used for the hotstore (eg LMDB). We plan to experiment with these options in the future.

Once the splitstore has been enabled, the existing monolithic blockstore becomes the coldstore. On the first head change notification, the splitstore will warm up the hotstore by copying all reachable objects from the current tipset into the hotstore.  All new writes go into the hotstore, with the splitstore tracking the write epoch. Once 5 finalities have elapsed, and every finality thereafter, the splitstore compacts by moving cold objects into the coldstore. There is also experimental support for garbage collection, whereby nunreachable objects are simply discarded.

## Activation

To enable the splitstore, add the following to `config.toml`:

```
[Chainstore]
  EnableSplitstore = true
```

## Experimentation

This feature is still experimental, so use at your own risk. Having said that, we would love if more adventurous users tried it out, and reported on their experience. In particular, we're curious to know about:

- any bugs or issues you stumble into
- the impact on block validation times when running with splitstore enabled
- the overall performance impact of enabling splitstore.

**Comment by vyzo** [comment]
So the first thing to do is just enable it! We'd love to hear back with your experience and help us improve it, as we want to make it the default in the future.

We also recommend applying this patch: https://github.com/filecoin-project/lotus/pull/5778, which should land to master (already approved) fairly soon.

**Reply by vyzo** [reply]
We have also opened issue https://github.com/filecoin-project/lotus/issues/5789 tracking next steps in development.

**Reply by vyzo** [reply]
Another update: We have an open PR (#5808) that does a few optimizations:
- no more full chain walking, just walk the state roots; we don't keep the entire chain hot any more
- reduce the compaciton threshold to 3 finalities; that means we keep only 2-3 (as opposed to 4-5) finalities in the hotstore.
- remove the full compaction, as it doesn't gain us anything; it was there to support GC, but we want to go a different direction with garbage collecting the coldstore according to configurable user policies.

**Comment by Fatman13** [comment]
Dummy question: What is finality in this context? Does it mean an epoch that has been finalized? How many epoch does it take for one to finalize? Thank you!

**Reply by vyzo** [reply]
A finality is 900 epochs; it's the maximum reorg depth.
Two finalities is the maximum back-reference depth.

**Comment by clinta** [comment]
I have been using splitstore for about a month now on a full node synced from genesis, it has been very helpful allowing me to put the hot store on fast NVME and the cold store on slower sata SSDs. I've seen no issues since enabling it.

**Comment by jennijuju** [comment]
> My "chain" is currently set to "splitstore".
But there is no storage capacity for "chain".
Can I delete it with rm -r.lotus/datastore/spiltstore/* and "restart lotus daemon"

you cant or you’ll lose all the recent chain data and have to restart from a snapshot again

**Comment by jennijuju** [comment]
further improvement: https://github.com/filecoin-project/lotus/issues/6577

**Reply by vyzo** [reply]
Just a heads up, #6474 brings great improvements -- in fact we call this splitstore v1 (what we currently have is v0).
We expect it to be in v1.11.1, together with the implementation of online gc for the coldstore -- see #6577.

Stay tuned!

**Comment by clinta** [comment]
I've run into an issue. I have limited disk space on my hot store, but lots of storage on my cold store. But due to a different error my lotus lost sync for a while. Now that lotus is trying to catch up, it never compacts and fills my hot store disk, then stops syncing again.

I tried starting lotus with --no-bootstrap in the hope that I could trigger a compaction manually, but there is no way to do that either.

**Reply by vyzo** [reply]
Ouch, this is indeed annoying, I would classify it as a BUG.

Can you open an issue for it so that we can track it and fix it?
I might be able to get to it tomorrow, so that we can include it in v1.11.1.

**Reply by vyzo** [reply]
As a (temporary workaround) solution for your current predicament, you can:
- disable the splitstore and delete the hotstore (really the splitstore directory)
- sync from snapshot into the coldstore
- reenable the splitstore

This should fix the system; I also recommend running latest master (soon to be in v1.11.1) as it brings in splitstore v1.

Sidenote: this solution, while generally applicable, might need a small tweak to force the splitstore to warmup: you need to delete `/splitstore/warmupEpoch` from the metadata store.
You don't have to do that if you are upgrading from v0 to v1, it will happen automatically.

**Reply by clinta** [reply]
Thanks, this is part of my ongoing effort to maintain a full node synced from genesis, not snapshot. So I'd rather not wipe out my coldstore with a snapshot.

In #6474 

> We change the compaction range model to be right to left, so that multiple compactions that would happen after a lengthy sync are coalesced to one.

Does this mean that it will not do any compaction until lotus is in sync?

**Reply by vyzo** [reply]
> Does this mean that it will not do any compaction until lotus is in sync?

Yes

> Thanks, this is part of my ongoing effort to maintain a full node synced from genesis, not snapshot. > So I'd rather not wipe out my coldstore with a snapshot.

Aaah, yes, let's not do that.
There is a solution, but might require you to write a bit of code as the tooling doesn't exist yet.

Basically the idea is:
- stop the node and disable splitstore
- copy the hotstore into the coldstore -- that's the part that will probably require some code, see #6737
- restart the node and let it sync
- reenable splitstore

In the meantime I'll try to fix the underlying issue by detecting when the node is way out of sync (or syncing from scratch) and redirecting writes to the coldstore for the sync, followed by a warmup

**Reply by vyzo** [reply]
Created issue #6769 for the bug.

**Comment by vyzo** [comment]
Some exciting news, as Splitstore.v1 is shipping with v1.11.1-rc1.
Lots of progress, best to check the README: https://github.com/filecoin-project/lotus/blob/master/blockstore/splitstore/README.md

**Comment by TippyFlitsUK** [comment]
I have been running SplitStore for a couple of weeks now and everything is looking very stable. My hot store is hovering between 100GiB and 150GiB which is ideal.
I have one question... Is it possible to specify an alternative path for my cold store. Both stores are currently located in my `.lotus` repo folder of my NVME. I would like to move my cold store to my long-term HDD storage cluster.
Many thanks!!

**Comment by vyzo** [comment]
yes, you can just move it and symlink!

**Comment by stuberman** [comment]
I added `[Chainstore] EnableSplitstore = true` in my `.lotus/config.toml` a few days ago.
I see the hot store growing, am I able to delete the `.lotus/datastore/chain` directory or contents as its size is not changing?

Current sizes
> 365G	.lotus/datastore/chain
> 60G	        .lotus/datastore/splitstore/hot.badger

**Comment by vyzo** [comment]
no, you cant safely delete the coldstore yet  -- you could use the discard
option however, and it wont gorw further.

**Reply by stuberman** [reply]
Thanks, I have plenty of extra space (1TiB) for now, looking forward to a feature that will compact the coldstore or remove older pieces.

I think we also need to understand how we would update the chain with a snapshot. I used to replace the chain every 3-4 weeks due to growth and performance. In the current model, would we still shut down the daemon, download and reimport a minimal snapshot to the .lotus/datastore/chain directory after deleting the old chain. I wish I had done that before enabling split store.
